juju/juju
# [dev] Refactor machine port types and cleanup API

## Description of change

This PR is part of the ongoing work for supporting per-endpoint (and by extension per-subnet) firewall rules. It consists primarily of refactoring preparatory work to ensure that the next chunks of work can land much easier.

The key changes from this PR are:
- The removal the `PortRange` type from state which is nothing more than a duplicate of the `core/network.PortRange` with a unit name attached. The state code that deals with machine port manipulation has been tweaked to use the types from core/network instead.
- The removal of the legacy `Port` type from `core/network`. The facades which return back individual port lists as opposed to port ranges have also been updated to generate `params.Port` lists by slicing the obtained `PortRange` values on the fly.
- The `Ports` type from state has been renamed to `machineSubnetPorts` to better convey its purpose and has been un-exported. In its place, we now use the `MachineSubnetPorts` interface to make it easier to write unit tests (see the changed facade tests)
- An auxiliary `UnitPorts` type has been added to make it easier to process a slice of `MachineSubnetPorts` (eventually one per subnet) and extract the opened port ranges for a unit across each subnet.
- The port query and manipulation methods have been renamed so they are consistent and un-ambiguous; any un-used of deprecated methods that were found have also been removed.

One important thing to be aware of is that our current modeling assumes that ports are opened across **all** subnets of a machine. As a result, none of the API methods provided by our facades have the means to break down port ranges by subnet in response payloads. To remain backwards-compatible with pre-2.9 facade clients, the aforementioned methods will fetch the port ranges by subnet and then lump them in a single sorted list (effectively dropping the subnet ID and de-duping them) before they are returned to the client. This works as expected as pre-2.9 clients don't have the means to open ports on distinct subnets anyway. 

However, to support 2.9 clients we will need to introduce new API methods or bump the facade versions and augment their payloads.


## QA steps

Bootstrap a controller and try various port-related operations (open/close/list ports etc.). Everything should work as before.
